doc/user/interop.md

```diff
@@ -100,6 +100,29 @@ objects.
 
 TruffleRuby ignores the `internal` flag to this message.
 
+### `KEY_INFO`
+
+If the object is a Ruby `Hash`, and the key is in the hash, `READABLE`,
+`WRITABLE` are set. If the key is not in the hash, neither are set. Note that
+the interface for `KEY_INFO` converts the key to a Java `String`, so if your
+keys are `Symbol`, they will not match.
```

Comment:
This is unclear. Do you mean that another language will pass a j.l.String which will be converted to a Ruby String, which will not match the Symbol key in the Hash?

Reply:
The problem is what should we do if you pass a Ruby string or symbol as the name to `Truffle::Interop.key_info`? I think we need to convert it to `j.l.String`, so that other languages can recognise it. Ideally we could keep the Ruby object and just allow other languages to recognise it as a string! In fact this is an excellent example of why we need this and I'll put that in the issue I have with cross-language strings. Is that clear?

Reply:
Is it specified that KEY_INFO should receive a j.l.String?
If the normal usage is a key from `KEYS` then the real issue is whether `KEYS` should return an array of `j.l.String` or an array of arbitrary-represented identifiers handled by `KEY_INFO` in that language.

If JavaSript calls KEY_INFO with a j.l.String, it seems there is no choice but to choose to interpret it as a Ruby Symbol or a Ruby String for the `Hash` case. Maybe we should just not support `j.l.String` to lookup a Hash key? It's again related to this namespace issue of separating methods, ivars, and Map-like interface.

Reply:
> Is it specified that KEY_INFO should receive a j.l.String?

It doesn't appear to be. But it is specified that `KEYS` should return `j.l.String`.

> Maybe we should just not support j.l.String to lookup a Hash key?

Perhaps. I'd rather change semantics in another PR, and just make what we have work here for now.

Reply:
Yes, I think the semantics of converting j.l.String to Ruby String is fine for a `Hash`.
Just the explanation needs to be a bit clearer. It should not talk about what `Truffle::Interop.key_info` does but how we respond to KEY_INFO when passed a Java String.

Reply:
Fixed.